hash/bucketer.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file contains the utilities to make bucketing decisions.
+
+package hash
+
+import (
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// BucketSet answers to what bucket does key X belong in a
+// consistent manner (consistent as in consistent hashing).
+// In addition Bucket implements reconciler.Bucket interface, so it
+// can be used both in leader election and in routing applications.
+type BucketSet struct {
+	// The name of this bucket. Required to use `Has`.
+	name string
+	// Stores the cached lookups. cache is internally thread safe.
+	cache *lru.Cache
+
+	// mu guards buckets.
+	mu sync.RWMutex
+	// All the bucket names. Needed for building hash universe.
+	// `name` must be in this set.
+	buckets sets.String
+}
+
+// Scientifically inferred preferred cache size.
+const cacheSize = 4096
+
+func newCache() *lru.Cache {
+	c, _ := lru.New(cacheSize)
+	return c
+}
+
+// NewBucketSet creates a new bucket set with the given name.
+func NewBucketSet(name string, bucketList sets.String) *BucketSet {
```

Comment:
I guess we can define a construct func with this func and `newStatefulSetConfig` for StatefulSet case in leaderelection package?

Reply:
Not sure I follow.

Reply:
I mean how we can use this in leaderelection package.

Reply:
Ah, yea in SS elector we can just call new and wrap the return